File: echo-post-bolt/project/ios/Runner/NativeVideoPlayerPlugin.swift
import Flutter
import UIKit
import AVFoundation

/// NativeVideoPlayerPlugin: AVPlayer-based single instance video player for iOS
/// 
/// Maintains one AVPlayer instance across all video switches to eliminate
/// memory leaks and buffer overflow issues. Uses conservative buffer settings
/// to keep memory under 50MB, mirroring the Android ExoPlayer approach.
public class NativeVideoPlayerPlugin: NSObject, FlutterPlugin {
    private var channel: FlutterMethodChannel!
    private var textureRegistry: FlutterTextureRegistry!
    
    // Single AVPlayer instance - never dispose, only reuse
    private var avPlayer: AVPlayer?
    private var playerLayer: AVPlayerLayer?
    private var textureRef: FlutterTextureRegistry?
    private var textureId: Int64?
    private var pixelBufferOutput: AVPlayerItemVideoOutput?
    private var displayLink: CADisplayLink?
    
    // Video texture for Flutter rendering
    private var videoTexture: VideoTexture?
    
    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(name: "echo_post/video_player", 
                                          binaryMessenger: registrar.messenger())
        let instance = NativeVideoPlayerPlugin()
        instance.channel = channel
        instance.textureRegistry = registrar.textures()
        registrar.addMethodCallDelegate(instance, channel: channel)
    }
    
    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        // Ensure all AVPlayer operations run on main thread
        DispatchQueue.main.async { [weak self] in
            switch call.method {
            case "initializePlayer":
                self?.initializePlayer(result: result)
            case "switchVideo":
                guard let args = call.arguments as? [String: Any],
                      let path = args["path"] as? String else {
                    result(FlutterError(code: "INVALID_ARGUMENT", 
                                      message: "Video path is required", 
                                      details: nil))
                    return
                }
                self?.switchVideo(path: path, result: result)
            case "play":
                self?.playVideo(result: result)
            case "pause":
                self?.pauseVideo(result: result)
            case "setVolume":
                guard let args = call.arguments as? [String: Any],
                      let volume = args["volume"] as? Double else {
                    result(FlutterError(code: "INVALID_ARGUMENT", 
                                      message: "Volume is required", 
                                      details: nil))
                    return
                }
                self?.setVolume(volume: volume, result: result)
            case "dispose":
                self?.disposePlayer(result: result)
            default:
                result(FlutterMethodNotImplemented)
            }
        }
    }
    
    private func initializePlayer(result: @escaping FlutterResult) {
        guard avPlayer == nil else {
            // Already initialized - return existing texture
            result(["textureId": textureId ?? -1])
            return
        }
        
        do {
            // Create single AVPlayer instance with optimized settings
            avPlayer = AVPlayer()
            
            // Enhanced: Conservative buffer settings for memory efficiency
            if let player = avPlayer {
                // Set preferred buffer size (similar to Android's buffer settings)
                player.automaticallyWaitsToMinimizeStalling = false
                
                // Create video texture for Flutter rendering
                videoTexture = VideoTexture()
                textureId = textureRegistry.register(videoTexture!)
                
                // Configure video output for texture rendering
                let pixelBufferAttributes: [String: Any] = [
                    kCVPixelBufferPixelFormatTypeKey as String: kCVPixelFormatType_32BGRA,
                    kCVPixelBufferWidthKey as String: 1920,
                    kCVPixelBufferHeightKey as String: 1080
                ]
                
                pixelBufferOutput = AVPlayerItemVideoOutput(pixelBufferAttributes: pixelBufferAttributes)
                
                // Setup display link for texture updates
                displayLink = CADisplayLink(target: self, selector: #selector(updateTexture))
                displayLink?.add(to: .main, forMode: .common)
                displayLink?.isPaused = true
                
                result(["textureId": textureId ?? -1])
            } else {
                result(FlutterError(code: "INIT_ERROR", 
                                  message: "Failed to create AVPlayer", 
                                  details: nil))
            }
        } catch {
            result(FlutterError(code: "INIT_ERROR", 
                              message: "Failed to initialize player: \(error.localizedDescription)", 
                              details: nil))
        }
    }
    
    private func switchVideo(path: String, result: @escaping FlutterResult) {
        guard let player = avPlayer else {
            result(FlutterError(code: "NOT_INITIALIZED", 
                              message: "Player not initialized", 
                              details: nil))
            return
        }
        
        // Enhanced: More aggressive cleanup before switch
        player.pause()
        player.replaceCurrentItem(with: nil)
        
        // Small delay to ensure cleanup completes (mirroring Android approach)
        DispatchQueue.main.asyncAfter(deadline: .now() + 0.05) {
            do {
                let url = URL(fileURLWithPath: path)
                guard FileManager.default.fileExists(atPath: path) else {
                    result(FlutterError(code: "FILE_NOT_FOUND", 
                                      message: "Video file does not exist: \(path)", 
                                      details: nil))
                    return
                }
                
                let asset = AVAsset(url: url)
                let playerItem = AVPlayerItem(asset: asset)
                
                // Add video output to new item
                if let output = self.pixelBufferOutput {
                    playerItem.add(output)
                }
                
                // Enhanced: Set conservative buffer settings
                playerItem.preferredForwardBufferDuration = 10.0 // 10s max buffer
                
                player.replaceCurrentItem(with: playerItem)
                
                // Start display link when video is ready
                self.displayLink?.isPaused = false
                
                result(true)
            } catch {
                result(FlutterError(code: "SWITCH_ERROR", 
                                  message: "Failed to switch video: \(error.localizedDescription)", 
                                  details: nil))
            }
        }
    }
    
    private func playVideo(result: @escaping FlutterResult) {
        guard let player = avPlayer else {
            result(FlutterError(code: "NOT_INITIALIZED", 
                              message: "Player not initialized", 
                              details: nil))
            return
        }
        
        player.play()
        displayLink?.isPaused = false
        result(nil)
    }
    
    private func pauseVideo(result: @escaping FlutterResult) {
        guard let player = avPlayer else {
            result(FlutterError(code: "NOT_INITIALIZED", 
                              message: "Player not initialized", 
                              details: nil))
            return
        }
        
        player.pause()
        displayLink?.isPaused = true
        result(nil)
    }
    
    private func setVolume(volume: Double, result: @escaping FlutterResult) {
        guard let player = avPlayer else {
            result(FlutterError(code: "NOT_INITIALIZED", 
                              message: "Player not initialized", 
                              details: nil))
            return
        }
        
        let clampedVolume = Float(max(0.0, min(1.0, volume)))
        player.volume = clampedVolume
        result(nil)
    }
    
    private func disposePlayer(result: @escaping FlutterResult) {
        cleanupResources()
        result(nil)
    }
    
    // Centralized resource cleanup
    private func cleanupResources() {
        displayLink?.invalidate()
        displayLink = nil
        
        avPlayer?.pause()
        avPlayer?.replaceCurrentItem(with: nil)
        avPlayer = nil
        
        if let textureId = textureId {
            textureRegistry.unregisterTexture(textureId)
        }
        
        videoTexture = nil
        textureId = nil
        pixelBufferOutput = nil
    }
    
    @objc private func updateTexture() {
        guard let output = pixelBufferOutput,
              let texture = videoTexture,
              let currentTime = avPlayer?.currentItem?.currentTime() else {
            return
        }
        
        if output.hasNewPixelBuffer(forItemTime: currentTime) {
            if let pixelBuffer = output.copyPixelBuffer(forItemTime: currentTime, itemTimeForDisplay: nil) {
                texture.updatePixelBuffer(pixelBuffer)
            }
        }
    }
}

/// Video texture implementation for Flutter rendering
class VideoTexture: NSObject, FlutterTexture {
    private var pixelBuffer: CVPixelBuffer?
    
    func copyPixelBuffer() -> Unmanaged<CVPixelBuffer>? {
        guard let pixelBuffer = pixelBuffer else { return nil }
        return Unmanaged.passRetained(pixelBuffer)
    }
    
    func updatePixelBuffer(_ buffer: CVPixelBuffer) {
        pixelBuffer = buffer
    }
} 

File: echo-post-bolt/project/android/app/src/main/kotlin/com/example/echopost/NativeVideoPlayerPlugin.kt
package com.example.echopost

import android.content.Context
import android.media.MediaMetadataRetriever
import android.os.Handler
import android.os.Looper
import android.view.Surface
import androidx.media3.common.MediaItem
import androidx.media3.exoplayer.DefaultLoadControl
import androidx.media3.exoplayer.ExoPlayer
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import io.flutter.view.TextureRegistry
import java.io.File

/// NativeVideoPlayerPlugin: ExoPlayer-based single instance video player
/// 
/// Maintains one ExoPlayer instance across all video switches to eliminate
/// memory leaks and buffer overflow issues from VideoPlayerController disposal.
/// Uses conservative buffer settings (3-10s) to keep memory under 50MB.
class NativeVideoPlayerPlugin : FlutterPlugin, MethodCallHandler {
    private lateinit var channel: MethodChannel
    private lateinit var textureRegistry: TextureRegistry
    private lateinit var context: Context
    
    // Single ExoPlayer instance - never dispose, only reuse
    private var exoPlayer: ExoPlayer? = null
    private var surface: Surface? = null
    private var textureEntry: TextureRegistry.SurfaceTextureEntry? = null
    
    // Main thread handler for ExoPlayer operations
    private val mainHandler = Handler(Looper.getMainLooper())
    
    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        channel = MethodChannel(flutterPluginBinding.binaryMessenger, "echo_post/video_player")
        channel.setMethodCallHandler(this)
        textureRegistry = flutterPluginBinding.textureRegistry
        // Fixed: Use application context to prevent memory leaks
        context = flutterPluginBinding.applicationContext
    }
    
    // Fixed: Proper cleanup on detach
    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
        
        // Clean up native resources on main thread
        mainHandler.post {
            cleanupResources()
        }
    }
    
    override fun onMethodCall(call: MethodCall, result: Result) {
        // Fixed: Ensure all ExoPlayer operations run on main thread
        mainHandler.post {
            when (call.method) {
                "initializePlayer" -> initializePlayer(result)
                "switchVideo" -> {
                    val path = call.argument<String>("path")
                    if (path != null) {
                        switchVideo(path, result)
                    } else {
                        result.error("INVALID_ARGUMENT", "Video path is required", null)
                    }
                }
                "pause" -> pauseVideo(result)
                "play" -> playVideo(result)
                "setVolume" -> {
                    val volume = call.argument<Double>("volume")
                    if (volume != null) {
                        setVolume(volume, result)
                    } else {
                        result.error("INVALID_ARGUMENT", "Volume is required", null)
                    }
                }
                "getVideoSize" -> {
                    val path = call.argument<String>("path")
                    if (path != null) {
                        getVideoSize(path, result)
                    } else {
                        result.error("INVALID_ARGUMENT", "Video path is required", null)
                    }
                }
                "dispose" -> disposePlayer(result)
                else -> result.notImplemented()
            }
        }
    }
    
    private fun initializePlayer(result: Result) {
        try {
            if (exoPlayer != null) {
                // Already initialized - return existing texture
                result.success(mapOf("textureId" to textureEntry?.id()))
                return
            }
            
            // Create texture for video rendering
            textureEntry = textureRegistry.createSurfaceTexture()
            surface = Surface(textureEntry!!.surfaceTexture())
            
            // Enhanced: More conservative buffer settings for memory efficiency
            val loadControl = DefaultLoadControl.Builder()
                .setBufferDurationsMs(
                    3_000,   // 3s min buffer (very conservative)
                    10_000,  // 10s max buffer (reduced from 15s)
                    1_500,   // 1.5s playback buffer
                    3_000    // 3s rebuffer
                )
                .build()
            
            // Create single ExoPlayer instance with optimized settings
            exoPlayer = ExoPlayer.Builder(context)
                .setLoadControl(loadControl)
                .build()
                .apply {
                    setVideoSurface(surface)
                    playWhenReady = false
                    volume = 1.0f
                }
            
            val textureId = textureEntry!!.id()
            result.success(mapOf("textureId" to textureId))
            
        } catch (e: Exception) {
            result.error("INIT_ERROR", "Failed to initialize player: ${e.message}", null)
        }
    }
    
    private fun switchVideo(path: String, result: Result) {
        val player = exoPlayer
        if (player == null) {
            result.error("NOT_INITIALIZED", "Player not initialized", null)
            return
        }
        
        try {
            // Enhanced: More aggressive cleanup before switch
            player.stop()
            player.clearMediaItems()
            
            // Small delay to ensure cleanup completes
            mainHandler.postDelayed({
                try {
                    val file = File(path)
                    if (!file.exists()) {
                        result.error("FILE_NOT_FOUND", "Video file does not exist: $path", null)
                        return@postDelayed
                    }
                    
                    val mediaItem = MediaItem.fromUri(file.toURI().toString())
                    player.setMediaItem(mediaItem)
                    player.prepare()
                    
                    result.success(true)
                } catch (e: Exception) {
                    result.error("SWITCH_ERROR", "Failed to switch video: ${e.message}", null)
                }
            }, 50) // 50ms delay for cleanup
            
        } catch (e: Exception) {
            result.error("SWITCH_ERROR", "Failed to prepare video switch: ${e.message}", null)
        }
    }
    
    private fun playVideo(result: Result) {
        val player = exoPlayer
        if (player == null) {
            result.error("NOT_INITIALIZED", "Player not initialized", null)
            return
        }
        
        try {
            player.play()
            result.success(null)
        } catch (e: Exception) {
            result.error("PLAY_ERROR", "Failed to play video: ${e.message}", null)
        }
    }
    
    private fun pauseVideo(result: Result) {
        val player = exoPlayer
        if (player == null) {
            result.error("NOT_INITIALIZED", "Player not initialized", null)
            return
        }
        
        try {
            player.pause()
            result.success(null)
        } catch (e: Exception) {
            result.error("PAUSE_ERROR", "Failed to pause video: ${e.message}", null)
        }
    }
    
    private fun setVolume(volume: Double, result: Result) {
        val player = exoPlayer
        if (player == null) {
            result.error("NOT_INITIALIZED", "Player not initialized", null)
            return
        }
        
        try {
            val clampedVolume = volume.coerceIn(0.0, 1.0).toFloat()
            player.volume = clampedVolume
            result.success(null)
        } catch (e: Exception) {
            result.error("VOLUME_ERROR", "Failed to set volume: ${e.message}", null)
        }
    }
    
    private fun getVideoSize(path: String, result: Result) {
        try {
            // Use MediaMetadataRetriever to read width/height
            val retriever = MediaMetadataRetriever().apply {
                setDataSource(path)
            }
            val width = retriever.extractMetadata(
                MediaMetadataRetriever.METADATA_KEY_VIDEO_WIDTH
            )!!.toInt()
            val height = retriever.extractMetadata(
                MediaMetadataRetriever.METADATA_KEY_VIDEO_HEIGHT
            )!!.toInt()
            retriever.release()

            result.success(mapOf("width" to width, "height" to height))
        } catch (e: Exception) {
            result.error("SIZE_ERROR", "Failed to get video size: ${e.message}", null)
        }
    }
    
    // Enhanced: Explicit disposal method for clean shutdown
    private fun disposePlayer(result: Result) {
        try {
            cleanupResources()
            result.success(null)
        } catch (e: Exception) {
            result.error("DISPOSE_ERROR", "Failed to dispose player: ${e.message}", null)
        }
    }
    
    // Centralized resource cleanup
    private fun cleanupResources() {
        try {
            exoPlayer?.release()
            surface?.release()
            textureEntry?.release()
            
            exoPlayer = null
            surface = null
            textureEntry = null
            
        } catch (e: Exception) {
            // Log error but don't crash during cleanup
            println("Warning: Error during resource cleanup: ${e.message}")
        }
    }
} 